File: darwin/pro_video_editor/Sources/pro_video_editor/src/shared/features/split/SplitVideo.swift
import AVFoundation
import Foundation

/// Service for frame-accurate splitting of a single video into two files.
///
/// Without an explicit bitrate each half is **stream-copied**
/// (`AVAssetExportPresetPassthrough`): no re-encode, so the source codec, frame
/// rate, bit depth and HDR metadata survive and the export is bounded by I/O
/// instead of encoder throughput. The cut stays frame-accurate because
/// AVFoundation copies from the preceding keyframe and gates playback start
/// with a container edit list — players that honor edit lists (AVPlayer,
/// ExoPlayer, ffmpeg) start exactly at the requested frame. A source that
/// cannot be stream-copied falls back to the re-encoding preset below.
///
/// With an explicit bitrate (or as that fallback) the half is re-encoded from
/// the exact split frame via the closest export preset. Either way this
/// deliberately does **not** use the rendering compositor or any effects
/// pipeline; it is just two trimmed `AVAssetExportSession` passes, which keeps
/// it fast and predictable.
///
/// The two halves are exported sequentially so only one decoder/encoder is
/// active at a time, and every export is guarded by two watchdogs: an inner
/// stall bound (``defaultStallTimeout`` — no forward progress for N seconds) and
/// an outer hard bound (``defaultExportTimeout``). Either one force-cancels the
/// session and fails with a diagnostic context describing *why* it stalled
/// (which half, last progress, segment/total durations, preset, audio). The job
/// therefore always terminates — success, failure, cancellation, stall or
/// timeout — and never leaves the method-channel result pending forever.
///
/// Note on stalls: a genuine hang here is almost always contention for the
/// shared VideoToolbox encode/decode session pool — the split runs on its own
/// `SplitVideoQueue` while speed renders run concurrently on `RenderVideoQueue`,
/// with no global cap on live `AVAssetExportSession`s. An exhausted pool leaves
/// a session at `progress == 0` indefinitely, which the stall bound now catches.
class SplitVideo {
  static let queue = DispatchQueue(label: "SplitVideoQueue")

  /// Default hard upper bound: the outer safety net. A single half may export
  /// this long before it is force-cancelled, even while still reporting
  /// progress. Overridable per-call via ``SplitVideoModel.exportTimeout``.
  static let defaultExportTimeout: TimeInterval = 120

  /// Default stall bound: if a half makes no forward progress for this long it
  /// is treated as stalled and force-cancelled. Overridable per-call via
  /// ``SplitVideoModel.stallTimeout``.
  static let defaultStallTimeout: TimeInterval = 12

  /// Starts an asynchronous frame-accurate split job.
  ///
  /// Writes `0 → splitUs` to `startOutputPath` and `splitUs → end` to
  /// `endOutputPath`, then reports the two paths via `onComplete`.
  @discardableResult
  static func split(
    inputPath: String,
    splitUs: Int64,
    startOutputPath: String,
    endOutputPath: String,
    outputFormat: String,
    bitrate: Int?,
    enableAudio: Bool,
    exportTimeout: TimeInterval = defaultExportTimeout,
    stallTimeout: TimeInterval = defaultStallTimeout,
    onProgress: @escaping (Double) -> Void,
    onComplete: @escaping ([String]) -> Void,
    onError: @escaping (Error) -> Void
  ) -> RenderJobHandle {
    let handle = RenderJobHandle()
    queue.async {
      let task = Task {
        do {
          guard FileManager.default.fileExists(atPath: inputPath) else {
            throw NSError(
              domain: "SplitVideo", code: 404,
              userInfo: [NSLocalizedDescriptionKey: "Input file not found: \(inputPath)"])
          }

          let asset = AVURLAsset(url: URL(fileURLWithPath: inputPath))
          let duration = try await loadDuration(of: asset)
          let splitTime = CMTime(value: splitUs, timescale: 1_000_000)

          guard splitTime > .zero, splitTime < duration else {
            throw NSError(
              domain: "SplitVideo", code: 1,
              userInfo: [
                NSLocalizedDescriptionKey:
                  "Split position \(splitUs)us is outside the video duration"
              ])
          }

          let reencodePreset = applyBitrate(requestedBitrate: bitrate)
          let fileType = mapFormatToMimeType(format: outputFormat)

          // Source for export: the original asset keeps audio + orientation;
          // for a silent split we re-wrap into a video-only composition.
          let exportAsset: AVAsset =
            enableAudio ? asset : try await videoOnlyAsset(from: asset, duration: duration)

          PluginLog.print(
            "✂️ Splitting at \(String(format: "%.3f", CMTimeGetSeconds(splitTime)))s "
              + "(\(outputFormat), \(bitrate == nil ? "passthrough" : "preset \(reencodePreset)"), "
              + "audio: \(enableAudio))")

          let totalSeconds = CMTimeGetSeconds(duration)
          let splitSeconds = CMTimeGetSeconds(splitTime)
          let endDuration = CMTimeSubtract(duration, splitTime)

          /// Exports one half, preferring the lossless passthrough fast path.
          ///
          /// Without an explicit bitrate the half is stream-copied first. A
          /// genuine export failure (e.g. a container/codec combination that
          /// cannot be copied) falls back to the re-encoding preset.
          /// Cancellations and watchdog stall/timeout errors propagate
          /// unchanged — retrying those would break cancel semantics and the
          /// wall-clock bounds the watchdog exists to enforce.
          func exportHalf(
            _ half: SplitExportDiagnostics.Half,
            timeRange: CMTimeRange,
            outputPath: String,
            progressBase: Double
          ) async throws {
            func attempt(_ preset: String) async throws {
              try await exportSegment(
                asset: exportAsset,
                timeRange: timeRange,
                outputPath: outputPath,
                fileType: fileType,
                preset: preset,
                diagnostics: SplitExportDiagnostics(
                  half: half, segmentSeconds: CMTimeGetSeconds(timeRange.duration),
                  splitSeconds: splitSeconds, totalSeconds: totalSeconds, preset: preset,
                  enableAudio: enableAudio),
                exportTimeout: exportTimeout,
                stallTimeout: stallTimeout,
                handle: handle,
                onProgress: { onProgress(progressBase + $0 * 0.5) })
            }

            guard bitrate == nil else { return try await attempt(reencodePreset) }
            do {
              try await attempt(AVAssetExportPresetPassthrough)
            } catch is CancellationError {
              throw CancellationError()
            } catch let error as NSError where error.domain == ExportWatchdog.errorDomain {
              throw error
            } catch {
              PluginLog.print(
                "✂️ Passthrough split (\(half.rawValue)) failed — falling back to "
                  + "\(reencodePreset): \(error.localizedDescription)")
              try Task.checkCancellation()
              try await attempt(reencodePreset)
            }
          }

          // First half: 0 → split.
          try await exportHalf(
            .start,
            timeRange: CMTimeRange(start: .zero, duration: splitTime),
            outputPath: startOutputPath,
            progressBase: 0)

          try Task.checkCancellation()

          // Second half: split → end.
          try await exportHalf(
            .end,
            timeRange: CMTimeRange(start: splitTime, duration: endDuration),
            outputPath: endOutputPath,
            progressBase: 0.5)

          onComplete([startOutputPath, endOutputPath])
        } catch {
          onError(error)
        }
      }
      handle.attach(task: task)
    }
    return handle
  }

  // MARK: - Helpers

  /// Exports a single time range of `asset` to `outputPath`, frame-accurate at
  /// the range start, guarded by the stall and hard-timeout bounds.
  private static func exportSegment(
    asset: AVAsset,
    timeRange: CMTimeRange,
    outputPath: String,
    fileType: AVFileType,
    preset: String,
    diagnostics: SplitExportDiagnostics,
    exportTimeout: TimeInterval,
    stallTimeout: TimeInterval,
    handle: RenderJobHandle,
    onProgress: @escaping (Double) -> Void
  ) async throws {
    let outputURL = URL(fileURLWithPath: outputPath)
    try? FileManager.default.createDirectory(
      at: outputURL.deletingLastPathComponent(), withIntermediateDirectories: true)
    // AVAssetExportSession refuses to overwrite an existing file.
    try? FileManager.default.removeItem(at: outputURL)

    guard let export = AVAssetExportSession(asset: asset, presetName: preset) else {
      throw NSError(
        domain: "SplitVideo", code: 3,
        userInfo: [NSLocalizedDescriptionKey: "Export session creation failed"])
    }

    export.outputURL = outputURL
    export.outputFileType = fileType
    // Both preset families honour `timeRange` frame-accurately: a re-encoding
    // preset decodes from the preceding keyframe and re-encodes from the exact
    // start; passthrough copies from the preceding keyframe and gates playback
    // start with a container edit list (verified: container duration is exact
    // to the microsecond even for a cut mid-way through an 8.4s GOP).
    export.timeRange = timeRange
    export.shouldOptimizeForNetworkUse = true

    // Attached before the gate wait below so a cancel arriving while the half
    // is still queued is honoured; it can only refuse the start from there,
    // never force-cancel a session that has yet to run.
    handle.attach(export: export)

    // Serialize against other encodes (split halves + concurrent renders) so
    // they don't starve each other on the hardware encoder. The wait is outside
    // the watchdog below, so queueing never counts as a stall.
    do {
      try await withExportSlot {
        try await runExportWithTimeout(
          export, diagnostics: diagnostics, exportTimeout: exportTimeout,
          stallTimeout: stallTimeout, handle: handle, onProgress: onProgress)
      }
    } catch {
      // The destination is the caller's own path, so a half that stopped
      // mid-write is indistinguishable from a finished one for anyone who just
      // checks that the file is there. Nothing downstream cleans it up — the
      // split reports the failure without saying how far each half got.
      try? FileManager.default.removeItem(at: outputURL)
      throw error
    }
  }

  /// Runs the export guarded by the shared ``ExportWatchdog`` (inner stall bound
  /// + outer hard bound). If a bound fires the session is force-cancelled and a
  /// diagnostic error is thrown; otherwise the monitor is torn down when the
  /// export finishes.
  private static func runExportWithTimeout(
    _ export: AVAssetExportSession,
    diagnostics: SplitExportDiagnostics,
    exportTimeout: TimeInterval,
    stallTimeout: TimeInterval,
    handle: RenderJobHandle,
    onProgress: @escaping (Double) -> Void
  ) async throws {
    try await ExportWatchdog.run(
      diagnostics: diagnostics,
      exportTimeout: exportTimeout,
      stallTimeout: stallTimeout,
      onProgress: onProgress,
      // A bound tight enough to fire while the session is still queued must not
      // cancel it — see `ExportSessionGuard.forceCancel`.
      cancel: { ExportSessionGuard.forceCancel(export) },
      body: { progress in
        try await ExportSessionDriver.run(
          export, handle: handle, label: "Split", failureDomain: "SplitVideo",
          onProgress: progress)
      })
  }

  /// Builds a video-only composition (dropping audio) while preserving the
  /// source orientation, used for silent splits.
  private static func videoOnlyAsset(from asset: AVAsset, duration: CMTime) async throws
    -> AVAsset
  {
    let composition = AVMutableComposition()
    let videoTrack = try await loadVideoTrack(from: asset)
    guard
      let compTrack = composition.addMutableTrack(
        withMediaType: .video, preferredTrackID: kCMPersistentTrackID_Invalid)
    else {
      throw NSError(
        domain: "SplitVideo", code: 2,
        userInfo: [NSLocalizedDescriptionKey: "Failed to create video track"])
    }
    try compTrack.insertTimeRange(
      CMTimeRange(start: .zero, duration: duration), of: videoTrack, at: .zero)

    if #available(iOS 15.0, macOS 13.0, *) {
      compTrack.preferredTransform = try await videoTrack.load(.preferredTransform)
    } else {
      compTrack.preferredTransform = videoTrack.preferredTransform
    }
    return composition
  }

  private static func loadDuration(of asset: AVAsset) async throws -> CMTime {
    if #available(iOS 15.0, macOS 13.0, *) {
      return try await asset.load(.duration)
    }
    return asset.duration
  }

  private static func loadVideoTrack(from asset: AVAsset) async throws -> AVAssetTrack {
    if #available(iOS 15.0, macOS 13.0, *) {
      let tracks = try await asset.loadTracks(withMediaType: .video)
      guard let track = tracks.first else {
        throw NSError(
          domain: "SplitVideo", code: 5,
          userInfo: [NSLocalizedDescriptionKey: "No video track found"])
      }
      return track
    } else {
      guard let track = asset.tracks(withMediaType: .video).first else {
        throw NSError(
          domain: "SplitVideo", code: 5,
          userInfo: [NSLocalizedDescriptionKey: "No video track found"])
      }
      return track
    }
  }
}

/// Immutable context for one exported half, used to enrich a stall/timeout
/// failure with actionable diagnostics. The bracketed suffix is intentionally
/// identical in structure to the Android side (`preset` here vs `mime` there).
struct SplitExportDiagnostics: ExportDiagnostics {
  enum Half: String { case start, end }

  let half: Half
  /// Duration of *this* segment in seconds.
  let segmentSeconds: Double
  /// Absolute split position in seconds.
  let splitSeconds: Double
  /// Total source duration in seconds.
  let totalSeconds: Double
  let preset: String
  let enableAudio: Bool

  /// e.g. `[half=end progress=0.00 segment=0.52s split=0.53s total=1.05s
  /// preset=AVAssetExportPresetHighestQuality audio=false]`
  func context(progress: Double) -> String {
    String(
      format:
        "[half=%@ progress=%.2f segment=%.2fs split=%.2fs total=%.2fs preset=%@ audio=%@]",
      half.rawValue, progress, segmentSeconds, splitSeconds, totalSeconds, preset,
      enableAudio ? "true" : "false")
  }

  func timeoutMessage(seconds: Int, progress: Double) -> String {
    "Split export timed out after \(seconds)s \(context(progress: progress))"
  }

  func stallMessage(seconds: Int, progress: Double) -> String {
    "Split export stalled after \(seconds)s with no progress \(context(progress: progress))"
  }
}

File: android/src/main/kotlin/ch/waio/pro_video_editor/src/features/split/SplitVideo.kt
package ch.waio.pro_video_editor.src.features.split

import mapFormatToMimeType
import android.content.Context
import android.media.MediaMetadataRetriever
import android.net.Uri
import android.os.Handler
import android.os.SystemClock
import androidx.media3.common.MediaItem
import androidx.media3.common.util.UnstableApi
import androidx.media3.transformer.Composition
import androidx.media3.transformer.EditedMediaItem
import androidx.media3.transformer.ExportException
import androidx.media3.transformer.ExportResult
import androidx.media3.transformer.ProgressHolder
import androidx.media3.transformer.Transformer
import ch.waio.pro_video_editor.src.features.render.helpers.ResilientVideoEncoderFactory
import ch.waio.pro_video_editor.src.features.render.models.RenderJobHandle
import ch.waio.pro_video_editor.src.shared.concurrency.ExportGate
import ch.waio.pro_video_editor.src.shared.logging.PluginLog as Log
import java.io.File
import java.util.concurrent.CancellationException
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicReference

/**
 * Frame-accurate splitting of a single video into two files.
 *
 * Each half is re-encoded from the exact split frame using a Media3
 * [Transformer] with a [MediaItem.ClippingConfiguration]. Because clipping at an
 * arbitrary (non-keyframe) position forces a transcode, the cut is
 * frame-accurate — unlike a transmux, which would snap to a keyframe.
 *
 * The two halves run sequentially (one encoder at a time) and every half is
 * guarded by two watchdogs: an inner stall bound (no forward progress for
 * `stallTimeoutMs`) and an outer hard bound (`exportTimeoutMs`). Either one
 * force-cancels the [Transformer] and fails with a diagnostic context (which
 * half, last progress, segment/total durations, mime, audio). Crucially,
 * [Transformer.cancel] does not invoke the listener, so cancellation, stall and
 * timeout deliver the terminal result explicitly through a single job-level
 * guard. The job therefore always terminates — success, error, cancellation,
 * stall or timeout — and never leaves the method-channel result pending.
 *
 * Note on stalls: a genuine hang here is almost always contention for the
 * device's limited hardware [android.media.MediaCodec] encoder pool — split and
 * concurrent speed renders both drive their [Transformer] on the main [Looper]
 * with no global cap on live encoder sessions. A second concurrent encoder can
 * block at `progress == 0`, which the stall bound now catches.
 */
@UnstableApi
class SplitVideo(private val context: Context) {
    companion object {
        private const val TAG = "SplitVideo"

        /** Default hard upper bound before a half is force-cancelled. */
        const val DEFAULT_EXPORT_TIMEOUT_MS = 120_000L

        /** Default stall bound: no forward progress for this long → stalled. */
        const val DEFAULT_STALL_TIMEOUT_MS = 12_000L

        private const val POLL_INTERVAL_MS = 200L
    }

    /**
     * Starts a frame-accurate split. Writes `0 → splitUs` to [startOutputPath]
     * and `splitUs → end` to [endOutputPath], reporting both paths via
     * [onComplete].
     */
    fun split(
        inputPath: String,
        splitUs: Long,
        startOutputPath: String,
        endOutputPath: String,
        outputFormat: String,
        bitrate: Int?,
        enableAudio: Boolean,
        mainHandler: Handler,
        exportTimeoutMs: Long = DEFAULT_EXPORT_TIMEOUT_MS,
        stallTimeoutMs: Long = DEFAULT_STALL_TIMEOUT_MS,
        onProgress: (Double) -> Unit,
        onComplete: (List<String>) -> Unit,
        onError: (Throwable) -> Unit
    ): RenderJobHandle {
        val finished = AtomicBoolean(false)
        val pollStop = AtomicBoolean(false)
        val transformerRef = AtomicReference<Transformer?>(null)
        val timeoutRef = AtomicReference<Runnable?>(null)
        // True once this job holds the process-wide encoder slot. Both halves
        // run under a single slot; it is released exactly once at job end.
        val gateHeld = AtomicBoolean(false)

        fun releaseGateIfHeld() {
            if (gateHeld.compareAndSet(true, false)) ExportGate.release()
        }

        fun disarmTimeout() {
            timeoutRef.getAndSet(null)?.let { mainHandler.removeCallbacks(it) }
        }

        fun finishSuccess(paths: List<String>) {
            if (!finished.compareAndSet(false, true)) return
            pollStop.set(true)
            disarmTimeout()
            releaseGateIfHeld()
            onProgress(1.0)
            onComplete(paths)
        }

        fun finishError(error: Throwable) {
            if (!finished.compareAndSet(false, true)) return
            pollStop.set(true)
            disarmTimeout()
            // Tear down the active encoder *before* releasing the gate, so the
            // next queued job can't start a second encoder while this one is
            // still live. finishError always runs on the main Looper (the thread
            // the Transformer was created on), so cancel synchronously.
            try {
                transformerRef.getAndSet(null)?.cancel()
            } catch (_: Exception) {
            }
            releaseGateIfHeld()
            onError(error)
        }

        val handle = RenderJobHandle {
            finishError(CancellationException("Split canceled"))
        }

        if (splitUs <= 0L) {
            finishError(IllegalArgumentException("splitUs must be greater than 0"))
            return handle
        }

        val inputFile = File(inputPath)
        if (!inputFile.exists()) {
            finishError(IllegalArgumentException("Input file not found: $inputPath"))
            return handle
        }

        mainHandler.post {
            if (finished.get()) return@post
            // Serialize against other encodes (concurrent renders/splits) so
            // they don't starve each other on the hardware encoder. The wait is
            // *before* exportHalf arms its stall watchdog, so queueing never
            // counts as a stall — a split behind a running render waits instead
            // of failing fast. onGranted always runs on the main Looper (either
            // synchronously here, or from another job's release()).
            ExportGate.acquire {
                gateHeld.set(true)
                // Cancelled/failed while queued: hand the slot straight back.
                if (finished.get()) {
                    releaseGateIfHeld()
                    return@acquire
                }
                // Half 1: 0 → split. On success, chain Half 2: split → end.
                exportHalf(
                    inputPath = inputPath,
                    half = "start",
                    startUs = 0L,
                    endUs = splitUs,
                    splitUs = splitUs,
                    outputPath = startOutputPath,
                    outputFormat = outputFormat,
                    bitrate = bitrate,
                    enableAudio = enableAudio,
                    mainHandler = mainHandler,
                    exportTimeoutMs = exportTimeoutMs,
                    stallTimeoutMs = stallTimeoutMs,
                    pollStop = pollStop,
                    transformerRef = transformerRef,
                    timeoutRef = timeoutRef,
                    onProgress = { onProgress(it * 0.5) },
                    onHalfComplete = {
                        if (!finished.get()) {
                            pollStop.set(false)
                            exportHalf(
                                inputPath = inputPath,
                                half = "end",
                                startUs = splitUs,
                                endUs = null,
                                splitUs = splitUs,
                                outputPath = endOutputPath,
                                outputFormat = outputFormat,
                                bitrate = bitrate,
                                enableAudio = enableAudio,
                                mainHandler = mainHandler,
                                exportTimeoutMs = exportTimeoutMs,
                                stallTimeoutMs = stallTimeoutMs,
                                pollStop = pollStop,
                                transformerRef = transformerRef,
                                timeoutRef = timeoutRef,
                                onProgress = { onProgress(0.5 + it * 0.5) },
                                onHalfComplete = {
                                    finishSuccess(listOf(startOutputPath, endOutputPath))
                                },
                                onHalfError = { finishError(it) }
                            )
                        }
                    },
                    onHalfError = { finishError(it) }
                )
            }
        }

        return handle
    }

    /**
     * Re-encodes a single clipped range to [outputPath]. Must be called on the
     * [mainHandler] thread (Media3 [Transformer] requires a Looper).
     *
     * [half] is `"start"` or `"end"` and [splitUs] the absolute split position;
     * both are used only to build the diagnostic context of a stall/timeout
     * failure. The source duration is probed lazily on the failure path so a
     * successful split incurs no extra work.
     */
    private fun exportHalf(
        inputPath: String,
        half: String,
        startUs: Long,
        endUs: Long?,
        splitUs: Long,
        outputPath: String,
        outputFormat: String,
        bitrate: Int?,
        enableAudio: Boolean,
        mainHandler: Handler,
        exportTimeoutMs: Long,
        stallTimeoutMs: Long,
        pollStop: AtomicBoolean,
        transformerRef: AtomicReference<Transformer?>,
        timeoutRef: AtomicReference<Runnable?>,
        onProgress: (Double) -> Unit,
        onHalfComplete: () -> Unit,
        onHalfError: (Throwable) -> Unit
    ) {
        val outputFile = File(outputPath)
        outputFile.parentFile?.mkdirs()
        if (outputFile.exists()) outputFile.delete()

        val mimeType = mapFormatToMimeType(outputFormat)
        val encoderFactory = ResilientVideoEncoderFactory(
            context = context,
            mimeType = mimeType,
            bitrate = bitrate,
        )

        // Frame-accurate clipping: a non-keyframe start forces Media3 to
        // transcode, cutting at the exact requested microsecond.
        val clipping = MediaItem.ClippingConfiguration.Builder()
            .setStartPositionUs(startUs)
        if (endUs != null) clipping.setEndPositionUs(endUs)

        val mediaItem = MediaItem.Builder()
            .setUri(Uri.fromFile(File(inputPath)))
            .setClippingConfiguration(clipping.build())
            .build()
        val editedMediaItem = EditedMediaItem.Builder(mediaItem)
            .setRemoveAudio(!enableAudio)
            .build()

        val halfDone = AtomicBoolean(false)
        // Last progress percentage (0..100) reported by the encoder, or -1 if
        // none yet. Read by the watchdogs to enrich the failure message.
        val lastPercent = java.util.concurrent.atomic.AtomicInteger(-1)

        val transformer = Transformer.Builder(context)
            .setEncoderFactory(encoderFactory)
            .setVideoMimeType(mimeType)
            .addListener(object : Transformer.Listener {
                override fun onCompleted(composition: Composition, result: ExportResult) {
                    if (!halfDone.compareAndSet(false, true)) return
                    pollStop.set(true)
                    timeoutRef.getAndSet(null)?.let { mainHandler.removeCallbacks(it) }
                    onProgress(1.0)
                    onHalfComplete()
                }

                override fun onError(
                    composition: Composition,
                    result: ExportResult,
                    exception: ExportException
                ) {
                    if (!halfDone.compareAndSet(false, true)) return
                    pollStop.set(true)
                    timeoutRef.getAndSet(null)?.let { mainHandler.removeCallbacks(it) }
                    outputFile.delete()
                    onHalfError(exception)
                }
            })
            .build()
        transformerRef.set(transformer)

        // Diagnostics are built only when a watchdog fires, so the source
        // duration probe never runs on a successful split.
        fun diagnostics(): SplitExportDiagnostics {
            val totalUs = probeDurationUs(inputPath)
            val segmentUs = when {
                endUs != null -> endUs - startUs
                totalUs > 0 -> totalUs - startUs
                else -> -1L
            }
            return SplitExportDiagnostics(
                half = half,
                segmentUs = segmentUs,
                splitUs = splitUs,
                totalUs = totalUs,
                mimeType = mimeType,
                enableAudio = enableAudio,
            )
        }

        fun lastFraction(): Double = lastPercent.get().coerceAtLeast(0) / 100.0

        // Single terminal path: force-cancel the (listener-silent) transformer,
        // clean up and surface [cause] exactly once. Used for a synchronous
        // start() failure whose cause is already known.
        fun failHalf(cause: Throwable) {
            if (!halfDone.compareAndSet(false, true)) return
            pollStop.set(true)
            timeoutRef.getAndSet(null)?.let { mainHandler.removeCallbacks(it) }
            try {
                transformer.cancel()
            } catch (_: Exception) {
            }
            outputFile.delete()
            onHalfError(cause)
        }

        // Watchdog terminal path: claim the terminal and tear the encoder down
        // synchronously (fast, main thread), then build the diagnostic message —
        // whose duration probe (MediaMetadataRetriever) is blocking file I/O — on
        // a background thread so it never stalls the UI thread on the already-
        // stressed failure path.
        fun failHalfDiagnosed(stall: Boolean, seconds: Long) {
            if (!halfDone.compareAndSet(false, true)) return
            pollStop.set(true)
            timeoutRef.getAndSet(null)?.let { mainHandler.removeCallbacks(it) }
            try {
                transformer.cancel()
            } catch (_: Exception) {
            }
            outputFile.delete()
            val progress = lastFraction()
            Thread {
                val diag = diagnostics()
                val message =
                    if (stall) diag.stallMessage(seconds, progress)
                    else diag.timeoutMessage(seconds, progress)
                Log.e(TAG, message)
                // Deliver back on the main Looper: onHalfError chains into the
                // job-level finish, which releases the encoder gate and may hand
                // the slot to the next queued job — that job's export must start
                // on the Looper thread, not this probe thread.
                mainHandler.post { onHalfError(IllegalStateException(message)) }
            }.start()
        }

        // Outer hard watchdog: a still-progressing but never-finishing export is
        // force-cancelled and surfaced as an error instead of hanging.
        val timeout = Runnable {
            if (halfDone.get()) return@Runnable
            failHalfDiagnosed(stall = false, seconds = Math.round(exportTimeoutMs / 1000.0))
        }
        timeoutRef.set(timeout)
        mainHandler.postDelayed(timeout, exportTimeoutMs)

        try {
            transformer.start(editedMediaItem, outputPath)
        } catch (e: Exception) {
            failHalf(e)
            return
        }

        // Progress polling + inner stall watchdog. `lastAdvanceAt` marks the
        // last time progress moved forward; if it never does for stallTimeoutMs
        // (the reported "stuck at 0%" case) the half is failed early.
        val progressHolder = ProgressHolder()
        mainHandler.post(object : Runnable {
            private var lastProgress = -1
            private var lastAdvanceAt = SystemClock.uptimeMillis()

            override fun run() {
                if (pollStop.get() || halfDone.get()) return
                transformer.getProgress(progressHolder)
                val now = SystemClock.uptimeMillis()
                val percent = progressHolder.progress
                if (percent >= 0) {
                    lastPercent.set(percent)
                    onProgress(percent / 100.0)
                    if (percent > lastProgress) {
                        lastProgress = percent
                        lastAdvanceAt = now
                    }
                }
                // Media3 reports progress in whole percent, so "forward
                // progress" only ticks per 1%. A healthy encode that advances
                // slower than 1% per stallTimeout would be misread as stalled;
                // in practice that means a single encode longer than ~100×
                // stallTimeout, which the outer hard bound already caps and
                // which the ExportGate now keeps free of pool contention.
                if (stallTimeoutMs > 0 && now - lastAdvanceAt >= stallTimeoutMs) {
                    failHalfDiagnosed(
                        stall = true,
                        seconds = Math.round(stallTimeoutMs / 1000.0),
                    )
                    return
                }
                mainHandler.postDelayed(this, POLL_INTERVAL_MS)
            }
        })
    }

    /** Best-effort source duration in microseconds, or -1 if it cannot be read. */
    private fun probeDurationUs(path: String): Long {
        val retriever = MediaMetadataRetriever()
        return try {
            retriever.setDataSource(path)
            val ms = retriever
                .extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION)
                ?.toLongOrNull() ?: 0L
            if (ms > 0L) ms * 1000L else -1L
        } catch (e: Exception) {
            Log.w(TAG, "Failed to probe duration for split diagnostics: ${e.message}")
            -1L
        } finally {
            try {
                retriever.release()
            } catch (_: Exception) {
            }
        }
    }
}
